src/diag.rs

```diff
@@ -427,6 +435,87 @@ impl Diagnostic {
+            &DefCkDuplicateEquality(ref tok, fst, snd) => (format!("Definition Check: Duplicate equality for typecode '{code}'", code = t(tok)).into(), vec![(
+                AnnotationType::Warning,
+                format!("This is the second equality introduced for typecode '{label}'. Each typecode must have a unique equality.", label = t(tok)).into(),
+                stmt,
+                snd,
+            ), (
+                AnnotationType::Note,
+                "Definition was previously provided here.".into(),
+                sset.statement_or_dummy(StatementAddress::new(fst.0, NO_STATEMENT)),
+                fst.1,
+            )]),
+            &DefCkMalformedDefinition(syntax) => ("Definition Check: Malformed definition".into(), vec![(
+                AnnotationType::Error,
+                "Malformed definition".into(),
+                stmt,
+                stmt.label_span(),
+            ), (
+                AnnotationType::Note,
+                "Syntax axiom defined here".into(),
+                sset.statement(syntax),
+                sset.statement(syntax).label_span(),
+            )]),
+            &DefCkMisplacedPrimitive(span) => ("Definition Check: Misplaced 'primitive' command".into(), vec![(
+                AnnotationType::Warning,
+                "there is no pending definition for this".into(),
+                stmt,
+                span,
+            )]),
+            &DefCkMalformedEquality(prev_saddr, tok) => ("Definition Check: Malformed equality or equality theorem".into(), vec![(
+                AnnotationType::Error,
+                "Malformed equality".into(),
+                stmt,
+                tok,
+            ), (
+                AnnotationType::Note,
+                "declaration here does not have the shape of an equality (theorem)".into(),
+                sset.statement(prev_saddr),
+                sset.statement(prev_saddr).label_span(),
+            )]),
+            DefCkMissingDefinition => {
+                notes = &["If this is intentional, consider adding a $( $j primitive ...; $) command"];
+                ("Definition Check: Missing definition or 'primitive'".into(), vec![(
+                    AnnotationType::Error,
+                    "this syntax axiom has no corresponding definition".into(),
+                    stmt,
+                    stmt.label_span(),
+                )])
+            },
+            DefCkSyntaxUsedBeforeDefinition(tok, saddr) => (format!("Definition Check: '{label}' used before definition", label = t(tok)).into(), vec![(
+                AnnotationType::Error,
+                format!("this expression contains an occurrence of '{label}'", label = t(tok)).into(),
+                stmt,
+                stmt.span(),
+            ), (
+                AnnotationType::Note,
+                "syntax declared here".into(),
+                sset.statement(*saddr),
+                sset.statement(*saddr).label_span(),
+            )]),
```

Comment:
The label "syntax before definition" suggests that there is a definition somewhere, but actually the definition might be completely missing. Maybe we could hint at that in the message?

Reply:
be my guest